src/k_smallest.rs

```diff
@@ -1,20 +1,92 @@
-use alloc::collections::BinaryHeap;
-use core::cmp::Ord;
+use alloc::vec::{Vec, IntoIter};
+use core::cmp::{Ordering, Reverse};
 
-pub(crate) fn k_smallest<T: Ord, I: Iterator<Item = T>>(mut iter: I, k: usize) -> BinaryHeap<T> {
-    if k == 0 { return BinaryHeap::new(); }
+/// Consumes a given iterator, returning the minimum elements in **ascending** order.
+pub(crate) fn k_smallest_general<T, I: Iterator<Item = T>>(
+    mut iter: I,
+    k: usize,
+    mut comparator: impl FnMut(&T, &T) -> Ordering,
+) -> IntoIter<T> {
+    if k == 0 {
+        return Vec::new().into_iter();
+    }
+    let mut storage: Vec<T> = iter.by_ref().take(k).collect();
+    let mut heap = &mut storage[..];
 
-    let mut heap = iter.by_ref().take(k).collect::<BinaryHeap<_>>();
+    // Rearrange the into a valid heap by reordering from the second-bottom-most layer up
+    // Slightly faster than ordering on each insert
+    // (But only by a factor of lg(k) and I'd love to hear of a use case where that matters)
+    for i in (0..(heap.len() / 2 + 1)).rev() {
+        sift_down(heap, &mut comparator, i);
+    }
 
-    iter.for_each(|i| {
-        debug_assert_eq!(heap.len(), k);
-        // Equivalent to heap.push(min(i, heap.pop())) but more efficient.
-        // This should be done with a single `.peek_mut().unwrap()` but
-        //  `PeekMut` sifts-down unconditionally on Rust 1.46.0 and prior.
-        if *heap.peek().unwrap() > i {
-            *heap.peek_mut().unwrap() = i;
+    if k == heap.len() {
+        // Nothing else needs done if we didn't fill the storage in the first place
+        // Also avoids unexpected behaviour with restartable iterators
+        for val in iter {
+            if comparator(&heap[0], &val) == Ordering::Greater {
+                heap[0] = val;
+                sift_down(heap, &mut comparator, 0);
+            }
         }
-    });
+    }
 
-    heap
+    while heap.len() > 1 {
+        let last_idx = heap.len() - 1;
+        heap.swap(0, last_idx);
+        // Leaves the length shorter than the number of elements
+        // so that sifting does not disturb already popped elements
+        heap = &mut heap[..last_idx];
+        sift_down(heap, &mut comparator, 0);
+    }
+
+    storage.into_iter()
+}
+
+pub(crate) fn reverse_cmp<T, F>(cmp: F) -> impl Fn(&T, &T) -> Ordering
+where
+    F: Fn(&T, &T) -> Ordering,
+{
+    move |a, b| cmp(b, a)
+}
+
+pub(crate) fn reverse_key<T, K, F>(f: F) -> impl Fn(&T) -> Reverse<K>
+where
+    F: Fn(&T) -> K,
+{
+    move |a| Reverse(f(a))
+}
+
+/// Sift the element currently at `origin` **away** from the root until it is properly ordered
+fn sift_down<T>(
```

Comment:
Please make this a local function within `k_smallest_general`, it reduces jumping through the code when reading it.